src/core/meshcop/border_agent.cpp

```diff
@@ -642,18 +675,117 @@ void BorderAgent::Stop(void)
 {
     VerifyOrExit(mState != kStateStopped);
 
+#if OPENTHREAD_CONFIG_BORDER_AGENT_EPHEMERAL_KEY_ENABLE
+    if (mUsingEphemeralKey)
+    {
+        mUsingEphemeralKey = false;
+        mEphemeralKeyTimer.Stop();
+        mEphemeralKeyTask.Post();
+    }
+#endif
+
     mTimer.Stop();
     Get<Tmf::SecureAgent>().Stop();
 
     mState        = kStateStopped;
     mUdpProxyPort = 0;
-
     LogInfo("Border Agent stopped");
 
 exit:
     return;
 }
 
+#if OPENTHREAD_CONFIG_BORDER_AGENT_EPHEMERAL_KEY_ENABLE
+
+Error BorderAgent::SetEphemeralKey(const char *aKeyString, uint32_t aTimeout, uint16_t aUdpPort)
+{
+    Error    error  = kErrorNone;
+    uint16_t length = StringLength(aKeyString, kMaxEphemeralKeyLength + 1);
+
+    VerifyOrExit(mState == kStateStarted, error = kErrorInvalidState);
+    VerifyOrExit((length >= kMinEphemeralKeyLength) && (length <= kMaxEphemeralKeyLength), error = kErrorInvalidArgs);
+
+    if (!mUsingEphemeralKey)
+    {
+        mOldUdpPort = GetUdpPort();
+    }
+
+    Stop();
+
+    mUsingEphemeralKey = true;
+    mEphemeralKeyTask.Post();
+
+    Start(aUdpPort, reinterpret_cast<const uint8_t *>(aKeyString), static_cast<uint8_t>(length));
```

Comment:
when start error for port already in use, following should be skipped and there should be a way to restore ba state to started and allow further ephemeral key set operation.


```
// log example when start error
Feb 28 03:46:49 affe4207839f otbr-agent[289]: 00:14:59.185 [D] Cli-----------: Input: ba ephemeralkey set Z10Xm1 300000 49155
Feb 28 03:46:49 affe4207839f otbr-agent[289]: 00:14:59.185 [I] SecTransport--: State: Open -> Closed
Feb 28 03:46:49 affe4207839f otbr-agent[289]: 00:14:59.185 [I] BorderAgent---: Border Agent stopped
Feb 28 03:46:49 affe4207839f otbr-agent[289]: 00:14:59.185 [I] SecTransport--: State: Closed -> Open
Feb 28 03:46:49 affe4207839f otbr-agent[289]: 00:14:59.185 [C] Platform------: Failed to bind UDP socket: Address already in use
Feb 28 03:46:49 affe4207839f otbr-agent[289]: 00:14:59.185 [W] BorderAgent---: Failed to start agent: Failed
Feb 28 03:46:49 affe4207839f otbr-agent[289]: 00:14:59.185 [I] BorderAgent---: Allow ephemeral key for 300000 msec on port 49155
Feb 28 03:46:49 affe4207839f otbr-agent[289]: 00:14:59.185 [D] Cli-----------: Output: Done
Feb 28 03:46:49 affe4207839f otbr-agent[289]: 00:14:59.185 [D] Cli-----------: Output: BorderAgent callback: Ephemeral key in-use, port:49155

Feb 28 04:04:41 affe4207839f otbr-agent[289]: 00:32:51.072 [D] Cli-----------: Input: ba state
Feb 28 04:04:41 affe4207839f otbr-agent[289]: 00:32:51.072 [D] Cli-----------: Output: Stopped
Feb 28 04:04:41 affe4207839f otbr-agent[289]: 00:32:51.072 [D] Cli-----------: Output: Done

Feb 28 04:06:11 affe4207839f otbr-agent[289]: 00:34:21.058 [D] Cli-----------: Input: ba ephemeralkey set W10X12
Feb 28 04:06:11 affe4207839f otbr-agent[289]: 00:34:21.058 [D] Cli-----------: Output: Error 13: InvalidState
```

Reply:
I think we can use `otBorderAgentClearEphemeralKey()` (or `ba ephemeralkey clear`) to clear state before setting the ephemeral key again.